agent/src/main/java/org/openjdk/jmc/agent/Agent.java

```diff
@@ -71,9 +75,44 @@
 	public static void premain(String agentArguments, Instrumentation instrumentation) {
 		printVersion();
 		getLogger().fine("Starting from premain"); //$NON-NLS-1$
+		openUnsafePackage(instrumentation);
 		initializeAgent(agentArguments, instrumentation);
 	}
 
+	/**
+	 * Allows the agent's module to access the Unsafe class
+	 */
+	private static void openUnsafePackage(Instrumentation instrumentation) {
+		if (VersionUtils.getFeatureVersion() >= 11) {
+			Method redefineModule = getRedefineModuleMethod();
+
+			try {
+				redefineModule.invoke(instrumentation, Object.class.getModule(), Collections.emptySet(), // extraReads
+						Collections.emptyMap(), // extraExports
+						Collections.singletonMap("jdk.internal.misc", Collections.singleton(Agent.class.getModule())), // extraOpens
+						Collections.emptySet(), // extraUses
+						Collections.emptyMap() // extraProvides
+				);
+			} catch (IllegalAccessException | IllegalArgumentException | InvocationTargetException e) {
+				System.out.println(
```

Comment:
Perhaps moving the openUnsafePackage method out of the Agent class might be nice to keep the agent class clean. Also, since you use the same pattern in the VersionUtils, perhaps some static printAndExit(Exception e, int errorCode, String message) too. And perhaps use different exit codes for failing to open up the package vs not being able to determine the version.

Perhaps you could have an AgentStartupUtils under the agent.impl package and move the VersionUtils methods there too? These are only used when starting the agent.

Reply:
Good ideas, will rework based on that. Re VersionUtils, there's similar demands for that elsewhere; seeing now I probably should merge this with `VersionResolver`.